### ISSUE TITLE
OpenMMTools MCMC as an OpenPathSampling engine

### ISSUE THREAD

**Issue by dwhswenson** [choderalab/openmmtools#518]
I'm writing an OpenPathSampling engine to wrap the OpenMMTools `MCMCSampler`. This will allow users to do path sampling of Monte Carlo "trajectories." This came out of a user question in openpathsampling/openpathsampling#1040, and current development of the engine is in openpathsampling/openpathsampling#1041.

Some questions for the OpenMMTools developers (thanks in advance for your advice!):

#### Setting `sampler_state`

When starting a path sampling trajectory, we'll need to set the `sampler_state` for our `MCMCSampler`to the shooting point. The two ways I see to implement this are:

1. Keep the same `MCMCSampler`, create a new `SamplerState`, and set `ops_engine._sampler.sampler_state = sampler_state`.
2. Completely create a new `MCMCSampler` each time we set the sample state.

My first implementation is based on 1, since this is more analogous with our OpenMM engine (one `app.Simulation`; we set the positions/velocities/box vectors). 

The questions: Are there any advantages/disadvantages of one approach over the other? Are there any longer-term changes expected that might make one preferable? Is it safe to assume that `MCMCSampler.sampler_state` can be treated as a mutable attribute?

#### Serialization

(Vaguely related to #517.)

We'll need to serialize everything from OpenMMTools as well. Is there a way to serialize directly to string? In the current draft, we're using `openmmtools.utils.serialize`, which creates a `dict`. This means that the serialization of `Quantity` objects is then handled by our internal methods for serializing them. I think it would be cleaner if we didn't mix our custom serialization rules with yours, which is why I'm hoping for a pure OpenMMTools way to serialize to string/bytes (as with OpenMM, where we serialize integrators/systems to XML).

**Comment by andrrizzi** [comment]
Hi @dwhswenson .

I think you can safely either assign a new `SamplerState` to `MCMCSampler` or even set `MCMCSampler.sampler_state.positions = ...`. The `SamplerState` is applied to the cached `Context` at each iteration so it modifying that at runtime does not have side effects.

Some context for serialization: The `openmmtools.utils.serialize/deserialize` methods simply add to/read from the serialized `dict` the full name of the class, similarly to what `pickle/unpickle` does. The `dict` format is useful when you want to preserve the type of the data, which helped us with some netcdf serialization optimizations.

The easier way I see to equip our objects with string serialization is maybe to add (through a `Serializable` mixin class) a `to_json` format that simply does `json.dumps(self.__getstate__())`. Would this work for your application?

**Comment by dwhswenson** [comment]
Thanks @andrrizzi!

> I think you can safely either assign a new `SamplerState` to `MCMCSampler` or even set `MCMCSampler.sampler_state.positions = ...`.

Yes, this is even better. Somehow, when I looked at the code before, I missed the properties for `positions` and only saw the internal API `_set_positions`. I'll switch to doing it this way, since it is more analogous to the OpenMM `context.setPositions` call we do.

> The easier way I see to equip our objects with string serialization is maybe to add (through a `Serializable` mixin class) a `to_json` format that simply does `json.dumps(self.__getstate__())`. Would this work for your application?

This would be perfect. The main thing is that `simtk.unit` quantities are not themselves JSON-serializable. We have code internal to OPS to do the same thing (which is why the code in openpathsampling/openpathsampling#1041 currently works). But I thought it might be a cleaner design to completely serialize the OpenMMTools objects within OpenMMTools.

Actually, rather than implementing JSON serialization of quantities here, the better solution might be to add it to `unit.Quantity` upstream. After all, the crux of this issue is that we don't want to mix custom serialization approaches. So why not have an official version in OpenMM? I might take a stab at that, since I've already solved the problem once in OPS.

**Comment by andrrizzi** [comment]
> Actually, rather than implementing JSON serialization of quantities here, the better solution might be to add it to `unit.Quantity` upstream.

I agree that would be an ideal solution. I seem to remember the `Quantity` object also implements `__getstate__` and `__setstate__` methods that serialize it into `dict`. I don't know if they currently support also string serialization.
